listenbrainz/db/dump.py

```diff
@@ -0,0 +1,323 @@
+        'user_id',
+        'artist',
+        'release',
+        'recording',
+        'last_updated',
+    ),
+    'statistics.artist': (
+        'id',
+        'msid',
+        'name',
+        'release',
+        'recording',
+        'listener',
+        'listen_count',
+        'last_updated',
+    ),
+    'statistics.release': (
+        'id',
+        'msid',
+        'name',
+        'recording',
+        'listener',
+        'listen_count',
+        'last_updated',
+    ),
+    'statistics.recording': (
+        'id',
+        'msid',
+        'name',
+        'listener',
+        'listen_count',
+        'last_updated',
+    ),
+    'data_dump': (
+        'id',
+        'created',
+    ),
+}
+
+
+PRIVATE_TABLES = [
+    '"user"',
+    'api_compat.session',
+    'api_compat.token',
+]
+
+STATS_TABLES = [
+    'statistics.user',
+    'statistics.release',
+    'statistics.artist',
+    'statistics.recording',
+]
+
+def dump_postgres_db(location, threads=None):
+    """ Create postgres database dump in the specified location
+
+        Arguments:
+            location: Directory where the final dump will be stored
+            threads: Maximal number of threads to run during compression
+
+        Returns:
+            Path to created dump.
+    """
+
+    logger.info('Beginning dump of PostgreSQL database...')
+    time_now = datetime.today()
+    dump_path = os.path.join(location, 'listenbrainz-dump-{time}'.format(time=time_now.strftime('%Y%m%d-%H%M%S')))
+    create_path(dump_path)
+    logger.info('dump path: %s', dump_path)
+
+
+    logger.info('Creating dump of private data...')
+    try:
+        private_dump = create_private_dump(dump_path, time_now, threads)
+    except IOError as e:
+        log_ioerrors(logger, e)
+        logger.info('Removing created files and giving up...')
+        shutil.rmtree(dump_path)
+        return
+    except Exception as e:
+        logger.error('Unable to create private db dump due to error %s', str(e))
+        logger.info('Removing created files and giving up...')
+        shutil.rmtree(dump_path)
+        return
+    logger.info('Dump of private data created at %s!', private_dump)
+
+    logger.info('Creating dump of stats data...')
+    try:
+        stats_dump = create_stats_dump(dump_path, time_now, threads)
+    except IOError as e:
+        log_ioerrors(logger, e)
+        logger.info('Removing created files and giving up...')
+        shutil.rmtree(dump_path)
+        return
+    except Exception as e:
+        logger.error('Unable to create statistics dump due to error %s', str(e))
+        logger.info('Removing created files and giving up...')
+        shutil.rmtree(dump_path)
+        return
+    logger.info('Dump of stats data created at %s!', stats_dump)
+
+
+    logger.info('Creating a new entry in the data_dump table...')
+    try:
+        dump_id = add_dump_entry()
+    except Exception as e:
+        logger.info('Error while adding dump entry: %s', str(e))
+        return
```

Comment:
Not sure if we should remove the created dump files if adding a new dump entry fails and start again, or if we should loop until the new dump entry gets added. I'm leaning towards looping until new dump entry gets added, because if we error here, most of the heavy work of the dumps has been done and shouldn't be wasted.